python/user/bot/user_bot.py

```diff
@@ -26,29 +43,79 @@ def main():
     logging.info(f'starting a the user_bot for party {party}')
     client = network.aio_party(party)
 
+    def expired(after: int, posted_at: int) -> bool:
+        return (datetime.datetime.fromtimestamp(posted_at) + datetime.timedelta(
+            seconds=after)) < datetime.datetime.now()
+
     @client.ledger_ready()
-    def bot_ready(_):
-        logging.info(f'user_bot for party {party} is ready')
+    async def bot_ready(_):
+        global message_heap
+        global archive_after
+
+        existing_messages = client.find_active(Chat.Message, {'sender': client.party})
+        message_heap = [Message(int(cdata['postedAt']), cid)
+                        for cid, cdata in existing_messages.items()]
+        heapq.heapify(message_heap)
```

Comment:
mm, the ACS is queried and then stored in an in memory heap?
the heap is a good match for the problem here, but it does mean that the bot is a singleton: modifications to the heap will not be reflected in the ACS, and vice versa.
is it possible for multiple instances of this bot to execute concurrently (because it seems like that would quickly lead to sadness)?

Reply:
Yea, this is under the assumption of you can only launch 1 public bot instance per user. (And is how current A-S enforces). It will result in noop if the same party wants to launch another bot that's already running.

Reply:
I think the bigger issue is less about multiple instances of this one bot running and more about the fact this bot appears likely to fail if anything else removes a contract it expects to be able to remove itself.

I highly suggest we fix this, because the failure modes are not intuitive, will take quite some time to manifest themselves in a production setting, and are fairly easy to stumble on by accident.

Reply:
@mschaef-da that make sense. I think a fetch by cid and some try catch should do.